### PR TITLE
notebooks: add GQL API boilerplate and a read resolver

### PR DESCRIPTION
**Pull request by novoselrok**
This PR adds all the necessary boilerplate needed for an enterprise GQL resolver for notebooks. It also adds a read resolver for fetching a single notebook.

### CODE REVIEW

**Comment by novoselrok** [comment]
@camdencheek The diff is large due to the GQL boilerplate and interfaces necessary to set up a new enterprise resolver. I recommend you review this commit-by-commit. You can skip the https://github.com/sourcegraph/sourcegraph/pull/28968/commits/1c7602f7df02081be1262d34d52ad775787bd9cd commit, because it just updates the calls to `NewSchema` with a nil param.

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+		},
+	}
+}
+
+func marshalNotebookID(notebookID int64) graphql.ID {
+	return relay.MarshalID("Notebook", notebookID)
+}
+
+func unmarshalNotebookID(id graphql.ID) (notebookID int64, err error) {
+	err = relay.UnmarshalSpec(id, &notebookID)
```
Comment (camdencheek) [inline enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:36]: Should we do any validation here that the ID type is "Notebook"?

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+
+func (r *notebookResolver) ViewerCanManage(ctx context.Context) bool {
+	user, err := backend.CurrentUser(ctx, r.db)
+	if err != nil {
+		return false
+	}
+	if user == nil {
+		return false
+	}
+	return user.ID == r.notebook.CreatorUserID
```
Comment (camdencheek) [inline enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:102]: Elsewhere, there was a note that if the notebook's `CreatorUserID` is `nil`, that means the user was deleted. Does that mean if a notebook is owned by a user who has been deleted, it is not manageable by anyone?

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+	return nil, false
+}
+
+type markdownBlockResolver struct {
+	// block.type == NotebookMarkdownBlockType
+	block notebooks.NotebookBlock
+}
+
+func (r *markdownBlockResolver) ID() string {
+	return r.block.ID
```
Comment (camdencheek) [inline enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:136]: Sorry if I missed this in a previous PR, but why is `NotebookBlock.ID` a string?

**Review by camdencheek — APPROVED** [review]
Looks great. A couple of questions inline, but this looks good to mere as-is

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go (hunk 3 above)
Comment (novoselrok) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:136]: I'm using UUID strings generated on the client as block IDs. Seemed easier than having to add additional logic to implement auto-incrementing number IDs.

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go (hunk 2 above)
Comment (novoselrok) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:102]: Unfortunately, yes. Public notebooks will still be viewable, but not editable. I'm not sure if there is a better system at the moment, except letting site-admins change owners. But that would be out of scope right now 🙁

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go (hunk 1 above)
Comment (novoselrok) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:43]: Yep, we should. I was kind of surprised to see we don't do really do it elsewhere 😅: https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+relay.unmarshalSpec%28&patternType=literal

File: enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go (hunk 3 above)
Comment (camdencheek) [reply enterprise/cmd/frontend/internal/notebooks/resolvers/resolvers.go:143]: Ah, gotcha. Makes sense to me. Thanks!

**Pull request merged**